---
title: "FASTQuick Summary Report"
output:
  html_document:
    toc: true
    theme: united
params:
  input: 
    value: x
  SVDPrefix: 
    value: x
  FASTQuickInstallDir: 
    value: x
---

```{r setup, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{r libraries}
library(ggplot2)
library(scales)
library(knitr)

multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  library(grid)

  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)

  numPlots = length(plots)

  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                     ncol = cols, nrow = ceiling(numPlots/cols))
  }

  if (numPlots==1) {
    print(plots[[1]])

  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))

    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))

      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}

findBump<-function(m,x,cntCol,numLimits){
  cnt=0
  DepthCntMax=max(m[,cntCol])
  pivot=which.max(m[,cntCol])
  prev=DepthCntMax
  for(i in pivot:1)
  {
    if(cnt==numLimits)
    {
     break
    }

    if(m[i,2]>prev*1.2)#turning points
    {
      cnt=cnt+1
    }

    prev=m[i,2]
    min=i
  }

  cnt=0
  prev=DepthCntMax
  #print(dim(m))
  for(i in pivot:length(m[,1]))
  {
    if(cnt==numLimits)
    {
      break
    }

    if(m[i,2]>prev*1.2)
     {
      cnt=cnt+1
    }

      prev=m[i,2]
      max=i
  }

  return(list(MAX=max,MIN=min))
}

create.DenDist=function(InsertTable,positionCol,countCol){
  start_pos=InsertTable[1,positionCol]
  count=InsertTable[1,countCol]
  NewT=c(-1,0)
  for(i in 1:length(InsertTable[,positionCol]))
  {
    if(InsertTable[i,positionCol]<start_pos+5)
    {
      count=count+InsertTable[i,countCol]
    }
    else
    {
      NewT=rbind(c(start_pos,count),NewT)
      start_pos=InsertTable[i,positionCol]
      count=InsertTable[i,countCol]
    }
  }
  return(NewT)
}
```

```{r args}
#args=commandArgs(trailingOnly = TRUE)
#input=args[1]
#SVDPrefix=args[2]
#FASTQuickInstallDir=args[3]
input="~/Downloads/FASTQuick_test/EmpQualityInvestigation/output/regular_size_output_HG00553_exome"
SVDPrefix="~/Downloads/FASTQuick_test/EmpQualityInvestigation/index/regular_size.FASTQuick.fa.bed.phase3.vcf.gz"
FASTQuickInstallDir="/Users/fanzhang/Dropbox/ClionProjects/FASTQuick/"
#input=params$input
#SVDPrefix=params$SVDPrefix
#FASTQuickInstallDir=params$FASTQuickInstallDir
#print("Arguments:")
#print(input)
#print(SVDPrefix)
#print(FASTQuickInstallDir)
```

```{r analysis, warning=FALSE}
par(mfrow=c(2,2))

mydata= read.table(paste(input,".DepthDist",sep=""),header=FALSE)
mydata=mydata[2:150,]
colnames(mydata)=c("Depth","SiteCount")
lmt=findBump(mydata,1,2,3)
q1=ggplot(mydata,aes(x=Depth,y=SiteCount))+geom_line(color="#00BFC4")+
  ggtitle("Depth Distribution")+coord_cartesian(xlim=c(mydata[lmt$MIN,1],mydata[lmt$MAX,1]))+ theme(plot.title = element_text(hjust = 0.5))


mydata= read.table(paste(input,".EmpCycleDist",sep=""),header=FALSE)
maxCycle = 0
for (i in 1:150)
{
  maxCycle = i
  if(mydata[i,3] == 0) break
}
mydata=mydata[1:maxCycle,]
colnames(mydata)=c("SequencingCycle","VariantCount","BaseCount","EmpiricalQuality","ReadCount")
q2=ggplot(mydata)+geom_line(aes(y=EmpiricalQuality,x=SequencingCycle),color="#00BFC4")+
  ggtitle("Sequencing Cycle V.S. Empirical Quality")+coord_cartesian(xlim=c(0,maxCycle))+ylim(0,45)+
  theme(plot.title = element_text(hjust = 0.5,size=10))


mydata=read.table(paste(input,".EmpRepDist",sep=""),header=FALSE)
mydata=mydata[1:40,]
colnames(mydata)=c("SequencingQuality","VariantCount","BaseCount","EmpiricalQuality")
mydata$BaseCount=mydata$BaseCount
q3=ggplot(mydata)+geom_line(aes(y=EmpiricalQuality,x=SequencingQuality),color="#00BFC4")+
  ggtitle("Sequencing Quality V.S. Empirical Quality")+geom_abline(intercept=0, slope=1,color="purple",linetype="dotted")+
  xlim(0,40)+ylim(0,40)+ theme(plot.title = element_text(hjust = 0.5,size=10))

q4=ggplot(mydata)+geom_line(aes(y=BaseCount,x=SequencingQuality),color="red",linetype="dotted")+
  ggtitle("Base Count Distribution")+xlim(0,40)+ theme(plot.title = element_text(hjust = 0.5))

mydata= read.table(paste(input,".GCDist",sep=""),header=FALSE)
mydata=mydata[2:100,]
colnames(mydata)=c("GC","dummy1","dummy2","AvgDepth")
q5=ggplot(mydata,aes(x=GC,y=AvgDepth))+geom_line(color="#00BFC4")+
  ggtitle("GC Distribution")+coord_cartesian(xlim=c(0,100))+
  geom_abline(intercept=1, slope=0,color="red",linetype="dotted")+ theme(plot.title = element_text(hjust = 0.5))


mydata= read.table(paste(input,".AdjustedInsertSizeDist",sep="",row.names=NULL),colClasses=c("numeric","numeric"),header=FALSE)
mydata=mydata[-1,]
NewTable=create.DenDist(mydata,1,2)
Adjust.Table=NewTable[order(NewTable[,1]),]
sumNewTable=sum(Adjust.Table[,2])
Adjust.Table[,2]=Adjust.Table[,2]/sumNewTable
Adjust.Table=data.frame(Adjust.Table,rep("AdjustedInsertSize",length(Adjust.Table[,1])),row.names=NULL)
colnames(Adjust.Table)=c("InsertSize","Frequency","Category")

mydata= read.table(paste(input,".RawInsertSizeDist",sep="",row.names=NULL),colClasses=c("numeric","numeric"),header=FALSE)
mydata=mydata[-1,]
NewTable=create.DenDist(mydata,1,2)
Raw.Table=NewTable[order(NewTable[,1]),]
sumNewTable=sum(Raw.Table[,2])
Raw.Table[,2]=Raw.Table[,2]/sumNewTable
Raw.Table=data.frame(Raw.Table,rep("RawInsertSize",length(Raw.Table[,1])),row.names=NULL)
colnames(Raw.Table)=c("InsertSize","Frequency","Category")

Combined.Table=rbind(Raw.Table,Adjust.Table)
Combined.Table$Category=as.factor(Combined.Table$Category)

lmt=findBump(Adjust.Table,1,2,3)
q6=ggplot(Combined.Table,aes(x=InsertSize,y=Frequency,colour=Category))+geom_line()+
  ggtitle("InsertSize Distribution")+coord_cartesian(xlim=c(ifelse(Adjust.Table[lmt$MIN,1]>100,100,Adjust.Table[lmt$MIN,1]),ifelse(Adjust.Table[lmt$MAX,1]<1000,1000,Adjust.Table[lmt$MAX,1])))+
  theme(plot.title = element_text(hjust = 0.5),legend.position=c(0.85,0.9),legend.key.size = unit(0.2, "cm"),legend.text =element_text(size=5), legend.title = element_text(size=5) )


mydata <- scan(paste(input,".Summary",sep=""), what="", sep="\n")
fileLen=length(mydata)-1#minus contamination line
ExpectedDepth=as.numeric(strsplit(strsplit(mydata[fileLen-13]," ")[[1]][5], "\\[")[[1]][1])
EstimatedDepth=as.numeric(strsplit(strsplit(mydata[fileLen-12]," ")[[1]][4], "\\[")[[1]][1])

AccessibleFraction=as.numeric(strsplit(strsplit(mydata[fileLen-11]," ")[[1]][8],"%")[[1]][1])
EstimatedQ20Depth=as.numeric(strsplit(mydata[fileLen-3],"\\s+|:")[[1]][8])
EstimatedQ30Depth=as.numeric(strsplit(mydata[fileLen-2],"\\s+|:")[[1]][8])

Q20BaseFraction=as.numeric(strsplit(mydata[fileLen-6],":")[[1]][2])
Q30BaseFraction=as.numeric(strsplit(mydata[fileLen-5],":")[[1]][2])
Depth1=as.numeric(strsplit(mydata[fileLen-9],":")[[1]][2])
Depth2=as.numeric(strsplit(mydata[fileLen-8],":")[[1]][2])
Depth5=as.numeric(strsplit(mydata[fileLen-7],":")[[1]][2])
Depth10=as.numeric(strsplit(mydata[fileLen-6],":")[[1]][2])

plotdata=c(EstimatedQ30Depth,EstimatedQ20Depth,EstimatedDepth,ExpectedDepth)
#print(plotdata)
plotname=c("EstimatedQ30Depth","EstimatedQ20Depth","EstimatedDepth","ExpectedDepth")
mydate=cbind(read.table(text=plotname),plotdata)
q7=ggplot(mydate,aes(x=mydate[,1],y=mydate[,2]))+ggtitle("Depth")+
  scale_x_discrete(limits=c("EstimatedQ30Depth","EstimatedQ20Depth","EstimatedDepth","ExpectedDepth"))+
  theme(axis.text.x = element_text(size=8,angle = 310, hjust = 0),axis.title.x=element_blank(),plot.title = element_text(hjust = 0.5))+
  geom_bar(stat="identity",fill="#00BFC4",alpha=0.5)+ylab('Depth')#+xlim(0,length(plotdata)+1)+ylim(0,max(plotdata))

plotdata2=c(Q20BaseFraction,Q30BaseFraction,Depth1,Depth2,Depth5,Depth10)
plotname2=c("Q20","Q30","Depth 1","Depth 2","Depth 5","Depth 10")
#mydata2=cbind(read.table(text=plotname2),plotdata2)
mydata2=data.frame(plotname2,plotdata2)
q8=ggplot(mydata2,aes(x=mydata2[,1],y=mydata2[,2]))+ggtitle("Summary")+
  scale_x_discrete(limits=c("Q20","Q30","Depth 1","Depth 2","Depth 5","Depth 10"))+
  geom_bar(stat="identity",fill="#00BFC4",alpha=0.5)+ylab('Fraction')+ 
  theme(axis.text.x = element_text(angle = 310, hjust = 0.3),axis.title.x=element_blank(),plot.title = element_text(hjust = 0.5))


#ancestry
alphaScale=scale_alpha_discrete(range = c(0.9,0.3),guide=FALSE)
sizeScale=scale_size(range=c(1.5,1),guide=FALSE)

#set 1000g color scale
colScale = scale_color_manual(values=c('ESN'='#FFCD00','GWD'='#FFB900','LWK'='#CC9933','MSL'='#E1B919','YRI'='#FFB933','ACB'='#FF9900','ASW'='#FF6600',
                                    'CLM'='#CC3333','MXL'='#E10033','PEL'='#FF0000','PUR'='#CC3300','CDX'='#339900','CHB'='#ADCD00','CHS'='#00FF00',
                                       'JPT'='#008B00','KHV'='#00CC33','CEU'='#0000FF','FIN'='#00C5CD','GBR'='#00EBFF','IBS'='#6495ED','TSI'='#00008B',
                                       'BEB'='#8B008B','GIH'='#9400D3','ITU'='#B03060','PJL'='#E11289','STU'='#FF00FF','AFR'='#FFCD33','AFR/AMR'='#FF9900',
                                       'AMR'='#FF3D3D','EAS'='#ADFF33','EUR'='#64EBFF','SAS'='#FF30FF','UserSample'='#000000'),
                              breaks=c('ESN','GWD','LWK','MSL','YRI','ACB','ASW','CLM','MXL','PEL','PUR','CDX','CHB','CHS','JPT','KHV','CEU','FIN','GBR',
                                       'IBS','TSI','BEB','GIH','ITU','PJL','STU','AFR','AFR/AMR','AMR','EAS','EUR','SAS','UserSample'))
#set 1000g coordinates
POP=read.table(paste0(FASTQuickInstallDir,"/resource/1000g.pop"),header = F)
RefCoord.1kg=read.table(paste0(SVDPrefix,".V"),header = F)
RefCoord.1kg=RefCoord.1kg[,1:3]
RefCoord.1kg['POP'] <- POP$V2[match(RefCoord.1kg$V1, POP$V1)]
colnames(RefCoord.1kg)=c("ID","PC1","PC2","POP")
RefCoord=RefCoord.1kg


#assuming the input target sample has format ID PC1 PC2
TargetSample=read.table(file=paste0(input,".Ancestry"),header=T)
TargetSample=data.frame(ID=c("IntendedSample"),PC1=c(TargetSample[1,3]), PC2=c(TargetSample[2,3]), POP=c("UserSample"))
colnames(TargetSample)=c("ID","PC1","PC2","POP")
#plot RefCoord as grey
#CombinedData=TargetSample
#q9=ggplot(data=RefCoord,aes(PC1,PC2))+geom_point(color="grey")+geom_point(data=CombinedData,aes(PC1,PC2,color=POP))+
#  colScale+alphaScale+sizeScale
#prepare dataset for plot
CombinedData=rbind(RefCoord,TargetSample)
#plot RefCoord as colorful
q10=ggplot()+geom_point(data=CombinedData,aes(PC1,PC2,color=POP), alpha=0.5)+#geom_text(data=CombinedData,aes(PC1,PC2,label=POP),size=1)+
colScale+alphaScale+sizeScale
```
## FASTQ File List
```{r FASTQ List}
fastq.tbl = read.csv(file=paste0(input,".FASTQ.csv"),header=T)
kable(fastq.tbl, caption="FASTQ List Table")
```

## Data Production by FASTQ file
```{r DataProduction}
sequence.tbl = read.csv(file=paste0(input,".Sequence.csv"),header=T)
kable(sequence.tbl, caption="Data Production Table")
```

## Depth Distribution
```{r Depth Distribution, fig.align = "center"}
multiplot(q1,q2,q3,q4, cols=2)
```

## Summary Statistics
```{r summary}
summary.tbl = read.table(file=paste0(input,".Summary"),header=T, sep=':')
kable(summary.tbl, caption="Summary Statistics")
```

## Summary Plot
```{r Misc, fig.align = "center"}
multiplot(q5,q7,q6,q8,cols=2)
```

## Genetic Ancestry Plot
```{r ancestry plot, fig.align = "center"}
#output file
#multiplot(q9,cols=1)
multiplot(q10,cols=1)
```