<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>Fastqa by Griffan</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>Fastqa</h1>
        <h2></h2>
        <a href="https://github.com/Griffan/FastqA" class="button"><small>View project on</small> GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h1>
<a name="introduction" class="anchor" href="#introduction"><span class="octicon octicon-link"></span></a>Introduction:</h1>

<p>FastqA is a software used for fast validation of fastq files in short time.
It rapidly map reads to selected region and generate a variety of quality control statistics. </p>

<h1>
<a name="usage" class="anchor" href="#usage"><span class="octicon octicon-link"></span></a>Usage:</h1>

<p>FastqA has multiple modes:
1. Generate selected region using user specified parameters, then do the following analysis.
Typical command line instruction:
FastqA --vcf hapmap_3.3.b37.vcf --dbsnp dbSNP.b137.00-All.vcf.gz --ref hs37d5.fa --flank_len 1000 --var_short 1000 --mask 20141007.all.strict_mask.fasta --fastq_1 read1.fastq --fastq_2 read2.fastq --bam_out debug.bam --cal_dup  --prefix shuffled.all</p>

<ol>
<li><p>Use pre-selected sites to analyze. FastqA also accept single end fastq as well as fastq file list as input.
Typical command line instruction:
FastqA --vcf pre-selected.sites.vcf --dbsnp dbSNP.b137.00-All.vcf.gz --ref hs37d5.fa --flank_len 1000 --var_short 1000 --fastq_1 read1.fastq  --bam_out debug.bam --cal_dup  --prefix shuffled.all</p></li>
<li><p>Use bam files along with corresponding selected regions to analyze.
Typical command line instruction:
FastqA --vcf pre-selected.sites.vcf --dbsnp dbSNP.b137.00-All.vcf.gz --ref hs37d5.fa --flank_len 1000 --var_short 1000  --bam_in debug.bam --cal_dup  --prefix shuffled.all</p></li>
</ol>

<h1>
<a name="faq" class="anchor" href="#faq"><span class="octicon octicon-link"></span></a>FAQ:</h1>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/Griffan/FastqA/zipball/master" class="button">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/Griffan/FastqA/tarball/master" class="button">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/Griffan/FastqA"></a> is maintained by <a href="https://github.com/Griffan">Griffan</a>.</p>

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the Architect theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

  
  </body>
</html>
