<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>FastPopCon by Griffan</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>FastPopCon</h1>
        <h2></h2>
        <a href="https://github.com/Griffan/FastPopCon" class="button"><small>View project on</small> GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h3>
<a id="name" class="anchor" href="#name" aria-hidden="true"><span class="octicon octicon-link"></span></a>NAME</h3>

<p>FastPopCon, a <strong>Fast</strong>q file based <strong>Pop</strong>ulation identification and <strong>Con</strong>tamination detection tool</p>

<h3>
<a id="contents" class="anchor" href="#contents" aria-hidden="true"><span class="octicon octicon-link"></span></a>CONTENTS</h3>

<h3>
<a id="synopsis" class="anchor" href="#synopsis" aria-hidden="true"><span class="octicon octicon-link"></span></a>SYNOPSIS</h3>

<pre><code>FastPopCon index --vcf hapmap.vcf --dbsnp 00-All.vcf.gz  --ref hs37d5.fa --flank_len 250 --var_short 9000 --flank_long_len 1000 --var_long 1000 --mask 20141007.all.strict_mask.fasta 

FastPopCon align --ref hs37d5.fa --fq_list NA12878.fq.list --bam_out --cal_dup --flank_len 250 --var_short 9000 --flank_long_len 1000 --var_long 1000  --I --t 2  --prefix NA12878 --frac_samp 1.0 

FastPopCon pop --UD resource/hapmap.dat.UD --PC resource/hapmap.dat.V --mu resource/hapmap.dat.mu --gl NA12878.likelihood --bed resource/choose.bed.post.bed.allele.bed

FastPopCon con --prefix NA12878
</code></pre>

<h3>
<a id="description" class="anchor" href="#description" aria-hidden="true"><span class="octicon octicon-link"></span></a>DESCRIPTION</h3>

<p>FastPopCon is short for <strong>Fast</strong>q file based <strong>Pop</strong>ulation identification and <strong>Con</strong>tamination detection tool. It is designed for fast quality control analysis of fastq files. It rapidly map reads to selected region and generate a variety of quality control statistics.</p>

<h3>
<a id="commands-and-options" class="anchor" href="#commands-and-options" aria-hidden="true"><span class="octicon octicon-link"></span></a>COMMANDS AND OPTIONS</h3>

<p><strong>index</strong>   </p>

<pre><code>FastPopCon index --vcf [hapmap site vcf] --dbsnp [dbsnp site vcf]  --ref [reference fasta]  --flank_len [250] --var_short [9000] --flank_long_len [1000] --var_long [1000] --mask [repeat_mask.fasta] 
</code></pre>

<p>Index database sequences, using known variant sites to anchor informative region.</p>

<pre><code>OPTIONS
--vcf   STR path of input hapmap site vcf file 
--dbsnp STR path of input dbsnp site vcf file
--ref   STR path of reference genome fasta file
--mask  STR path of repetitive region  mask fasta file
--flank_len INT flanking region length of short-flanking-region variant
--var_short INT number of short-flanking-region variant
--flank_long_len    INT flanking region length of long-flanking-region variant
--var_long  INT number of long-flanking-region variant
</code></pre>

<p><strong>align</strong></p>

<pre><code>FastPopCon align --ref [reference genome fasta file] --fq_list [sample’s fastq list file] [--bam_out] [--cal_dup] --flank_len [250] --var_short [9000] --flank_long_len [1000] --var_long [1000]  [--I] --t [2]  --prefix [NA12878] --frac_samp [1.0]
</code></pre>

<p>Align short reads 70~300 bp to selected reference region to generate comprehensive quality control related statistics in very short time.</p>

<pre><code>OPTIONS
--ref   STR path of reference genome fasta file
--fq_list   STR path of fastq file list, format: [path of pair-end 1]\t[path of pair-end]
--fastq_1   STR path of pair end 1 fastq file
--fastq_2   STR path of pair end 2 fastq file
--bam_in    STR path of already aligned bam file
--bam_out   BOOL    output is bam format or not
--prefix    STR prefix of variety of output file associated with specific sample
--flank_len INT flanking region length of short-flanking-region variant
--var_short INT number of short-flanking-region variant
--flank_long_len    INT flanking region length of long-flanking-region variant
--var_long  INT number of long-flanking-region variant
--n float   Maximum edit distance if the value is INT, or the fraction of missing alignments given 2% uniform base error rate if FLOAT. In the latter case, the maximum edit distance is automatically chosen for different read lengths. [0.04]
--o INT Maximum number of gap opens [1]
--e INT Maximum number of gap extensions, -1 for k-difference mode (disallowing long gaps) [-1]
--i INT Disallow an indel within INT bp towards the ends [5]
--d INT Disallow a long deletion within INT bp towards the 3’-end [16]
--l INT Take the first INT subsequence as seed. If INT is larger than the query sequence, seeding will be disabled. For long reads, this option is typically ranged from 25 to 35 for ‘-k 2’. [inf]
--k INT Maximum edit distance in the seed [2]
--m INT Maximum entries in the queue [2000000]
--t INT Number of threads [1]
--R INT Stop searching when there are &gt;INT equally best hits [30]
--q INT Quality threshold for read trimming down to 35bp [0]
--RG    STR Read group name
--N BOOL    Non-iterative mode: search for all n-difference hits
--I BOOL    The input is in the Illumina 1.3+ read format (quality equals ASCII-64)
--L BOOL    Log-scaled gap penalty for long deletions 
--max_isize INT Maximal insert size for a read pair to be considered being mapped properly.[500] 
--max_occ   INT Maximum occurrences of a read for pairing. A read with more occurrences will be treated as a single-end read. Reducing this parameter helps faster pairing. [100000]
--is_sw BOOL    Enable Smith-Waterman for the unmapped mate.[True]
--n_multi   INT Maximum hits to output for paired reads.[3]
--N_multi   INT Maximum hits to output for discordant pairs.[10]
--ap_prior  FLOAT   Prior of chimeric rate (lower bound)[1.0e-05]
--force_isize   BOOL    Disable insert size estimate.[False]
--cal_dup   BOOL    Calculate PCR duplicated reads in all the statistics.[False]
--frac_samp FLOAT   Overall reads downsampling rate.[1]
</code></pre>

<p><strong>pop</strong></p>

<pre><code>FastPopCon pop --UD [resource/hapmap.dat.UD] --PC [resource/hapmap.dat.V] --mu [resource/hapmap.dat.mu] --gl [prefix.likelihood] --bed [resource/choose.bed.post.bed.allele.bed]
</code></pre>

<p>Identify individual’s population identity, ancestry information. The geometric distance in plot represents how close the relatedness is.</p>

<pre><code>OPTIONS
--UD    STR Path of UD matrix in resource directory
--PC    STR Path of V matrix in resource directory
--mu    STR Path of mu matrix in resource directory
--gl    STR Path of output likelihood file in align step, with prefix specified in align step and suffix likelihood.
--bed   STR Bed format file that specified markers used in pop inference, also can be found in resource directory.
</code></pre>

<p><strong>con</strong></p>

<pre><code>FastPopCon con --prefix [NA12878]
</code></pre>

<p>Estimate the probability that this sample is contaminated with other genomic material.</p>

<pre><code>OPTIONS
--prefix STR Specify the prefix used in previous steps, which will be used to retrieve all the information needed in this step.
</code></pre>

<h3>
<a id="examples" class="anchor" href="#examples" aria-hidden="true"><span class="octicon octicon-link"></span></a>EXAMPLES</h3>

<p>Some examples of common usage.</p>

<pre><code>src/FastqA index --vcf new_sites/All.hapmap.omni.HDGP.recode.vcf --dbsnp dbSNP/b137/00-All.vcf.gz  --ref hs37d5.fa --flank_len 250 --var_short 9000 --flank_long_len 1000 --var_long 1000 --mask 20141007.all.strict_mask.fasta

src/FastqA align --ref hs37d5.fa --fq_list NA12878.fq.list --bam_out --cal_dup --flank_len 250 --var_short 9000 --flank_long_len 1000 --var_long 1000  --I --t 2  --prefix NA12878 --frac_samp 1
</code></pre>

<h3>
<a id="see-also" class="anchor" href="#see-also" aria-hidden="true"><span class="octicon octicon-link"></span></a>SEE ALSO</h3>

<p>A list of related commands or functions.</p>

<h3>
<a id="bugs" class="anchor" href="#bugs" aria-hidden="true"><span class="octicon octicon-link"></span></a>BUGS</h3>

<p>List known bugs.</p>

<h3>
<a id="author" class="anchor" href="#author" aria-hidden="true"><span class="octicon octicon-link"></span></a>AUTHOR</h3>

<p>Fan Zhang (email:<a href="mailto:fanzhang@umich.edu">fanzhang@umich.edu</a>)</p>

<h3>
<a id="copyright" class="anchor" href="#copyright" aria-hidden="true"><span class="octicon octicon-link"></span></a>COPYRIGHT</h3>

<p>The full FastPopCon package is distributed under GPLv3.</p>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/Griffan/FastPopCon/zipball/master" class="button">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/Griffan/FastPopCon/tarball/master" class="button">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/Griffan/FastPopCon"></a> is maintained by <a href="https://github.com/Griffan">Griffan</a>.</p>

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the Architect theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

            <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-56053249-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
